RecoTracker/ExoRegionalTracking/plugins/ExoRegionProducer.cc

```diff
@@ -0,0 +1,231 @@
+      return (!a.entities().first->valid() || !a.entities().second->valid());
+    };
+    const auto pseudoROIPred = [](const ExoRegion &a) { return !a.valid(); };
+    distances.remove_if(distancePred);
+    pseudoROIs.remove_if(pseudoROIPred);
+  }
+
+  // Remove invalid ROIs.
+  const auto roiPred = [&](const ExoRegion &roi) {
+    if (!roi.valid())
+      return true;
+    const math::XYZVector x(roi.vx(), roi.vy(), roi.vz());
+    if ((minRadius_ >= 0.0 && (x - bs).rho() < minRadius_) ||
+        (genMatch_ != "None" && !roiContainsDecayVertex(x, decayVertices, roi.rParam())))
+      return true;
+    const double discriminatorValue = ((discriminatorCut_ > 0.0) ? getDiscriminatorValue(roi, *beamSpot) : 1.0);
+    if (discriminatorValue < discriminatorCut_)
+      return true;
+    return false;
+  };
+  pseudoROIs.remove_if(roiPred);
+
+  auto regionsOfInterest = make_unique<vector<reco::Vertex> >();
+  regionsOfInterest->clear();
+
+  const vector<double> error = {1.0, 0.0, 1.0, 0.0, 0.0, 1.0};
+
+  for (const auto &roi : pseudoROIs)
+    regionsOfInterest->emplace_back(reco::Vertex::Point(roi.vx(), roi.vy(), roi.vz()),
+                                    reco::Vertex::Error(error.begin(), error.end(), true, true));
+
+  event.put(move(regionsOfInterest));
+}
+
+void ExoRegionProducer::getDecayVertices(const edm::View<reco::GenParticle> &genParticles,
+                                         vector<math::XYZVector> &decayVertices) const {
+  for (const auto &genParticle : genParticles) {
+    if (genMatch_ == "HToSSTobbbb") {
+      if (genParticle.pdgId() != 5)
+        continue;
+      if (!genParticle.isHardProcess())
+        continue;
+      if (abs(genParticle.mother()->pdgId()) != 9000006)
+        continue;
+    } else if (genMatch_ == "stopToBottom") {
+      if (abs(genParticle.pdgId()) != 5)
+        continue;
+      if (!genParticle.isHardProcess())
+        continue;
+      if (abs(genParticle.mother()->pdgId()) < 1000000)
+        continue;
+    }
+
+    decayVertices.emplace_back(genParticle.vx(), genParticle.vy(), genParticle.vz());
+  }
+}
+
+const bool ExoRegionProducer::roiContainsDecayVertex(const math::XYZVector &roi,
+                                                     const vector<math::XYZVector> &decayVertices,
+                                                     const double r) const {
+  for (const auto &decayVertex : decayVertices)
+    if ((roi - decayVertex).r() < r)
+      return true;
+  return false;
+}
+
+const double ExoRegionProducer::getDiscriminatorValue(const ExoRegion &roi, const reco::BeamSpot &bs) const {
+  tensorflow::Tensor vertexTensor(tensorflow::DT_FLOAT, tensorflow::TensorShape({1, 40, 23}));
+  auto vertex_map = vertexTensor.tensor<float, 3>();
+  tensorflow::Tensor annulusTensor(tensorflow::DT_FLOAT, tensorflow::TensorShape({1, 10, 8}));
+  auto annulus_map = annulusTensor.tensor<float, 3>();
+
+  for (int i = 0, map_i = 0; map_i < 40; i++, map_i++) {
```

Comment:
wouldn't be better to have some `constexpr` for handling these "magic numbers" ?

Reply:
I'm familiar with `constexpr`, but I'm not sure how I could use it to handle the magic numbers in a clear way.